diff --git a/hw/core/machine.c b/hw/core/machine.c
index 7a13e9f014..914536f831 100644
--- a/hw/core/machine.c
+++ b/hw/core/machine.c
@@ -18,6 +18,7 @@
 #include "qapi/error.h"
 #include "qapi/qapi-visit-machine.h"
 #include "qemu/madvise.h"
+#include "qemu/config-file.h"
 #include "qom/object_interfaces.h"
 #include "sysemu/cpus.h"
 #include "sysemu/sysemu.h"
@@ -999,6 +1000,35 @@ void machine_add_audiodev_property(MachineClass *mc)
                                           "Audiodev to use for default machine devices");
 }
 
+static int vcpu_parse(void *opaque, QemuOpts *opts, Error **errp)
+{
+    MachineState *ms = opaque;
+    MachineClass *mc = MACHINE_GET_CLASS(ms);
+    if (opts) {
+        unsigned vcpu = qemu_opt_get_number(opts, "vcpunum", 0);
+        unsigned affinity = qemu_opt_get_number(opts,"affinity", 0);
+        if (vcpu < ms->smp.max_cpus) {
+            if (mc->vcpu_affinity[vcpu] == -1) {
+                mc->vcpu_affinity[vcpu] = affinity;
+            }
+            else {
+                error_setg(errp, "Duplicate affinity statement for vcpu %d", vcpu);
+                return -1;
+            }
+        }
+        else {
+            error_setg(errp, "VCPU %d exceeds maximum allowed (%d)", vcpu, ms->smp.max_cpus);
+            return -1;
+        }
+    }
+    return 0;
+}
+
+void parse_vcpu_opts(MachineState *ms)
+{
+    qemu_opts_foreach(qemu_find_opts("vcpu-opts"), vcpu_parse, ms, &error_fatal);
+}
+
 static void machine_class_init(ObjectClass *oc, void *data)
 {
     MachineClass *mc = MACHINE_CLASS(oc);
@@ -1013,6 +1043,9 @@ static void machine_class_init(ObjectClass *oc, void *data)
      */
     mc->smbios_memory_device_size = 2047 * TiB;
 
+    for (int i = 0; i < CPU_SETSIZE; i++)
+        mc->vcpu_affinity[i] = -1;
+
     /* numa node memory size aligned on 8MB by default.
      * On Linux, each node's border has to be 8MB aligned
      */
diff --git a/include/hw/boards.h b/include/hw/boards.h
index 5cddeb7fcb..e55641e596 100644
--- a/include/hw/boards.h
+++ b/include/hw/boards.h
@@ -43,6 +43,7 @@ void machine_set_cpu_numa_node(MachineState *machine,
                                Error **errp);
 void machine_parse_smp_config(MachineState *ms,
                               const SMPConfiguration *config, Error **errp);
+void parse_vcpu_opts(MachineState *ms);
 unsigned int machine_topo_get_cores_per_socket(const MachineState *ms);
 unsigned int machine_topo_get_threads_per_socket(const MachineState *ms);
 void machine_memory_devices_init(MachineState *ms, hwaddr base, uint64_t size);
@@ -264,6 +265,7 @@ struct MachineClass {
     int max_cpus;
     int min_cpus;
     int default_cpus;
+    int vcpu_affinity[CPU_SETSIZE];
     unsigned int no_serial:1,
         no_parallel:1,
         no_floppy:1,
diff --git a/qemu-options.hx b/qemu-options.hx
index 7fdc944965..004814f4de 100644
--- a/qemu-options.hx
+++ b/qemu-options.hx
@@ -383,6 +383,16 @@ SRST
     to guest if it's explicitly specified in -smp.
 ERST
 
+DEF("vcpu", HAS_ARG, QEMU_OPTION_vcpu,
+    "-vcpu [vcpunum=]n[,affinity=affinity]\n"
+    "                the index of the vCPU to pin\n"
+    "                affinity= the host CPU index to pin to\n",
+        QEMU_ARCH_ALL)
+SRST
+``-vcpu [vcpunum=]n[,affinity=affinity]``
+    VCPU Affinity. If specified, specify for all the CPUs.
+ERST
+
 DEF("numa", HAS_ARG, QEMU_OPTION_numa,
     "-numa node[,mem=size][,cpus=firstcpu[-lastcpu]][,nodeid=node][,initiator=node]\n"
     "-numa node[,memdev=id][,cpus=firstcpu[-lastcpu]][,nodeid=node][,initiator=node]\n"
diff --git a/system/cpus.c b/system/cpus.c
index 1c818ff682..875c19d948 100644
--- a/system/cpus.c
+++ b/system/cpus.c
@@ -664,7 +664,10 @@ const AccelOpsClass *cpus_get_accel(void)
 
 void qemu_init_vcpu(CPUState *cpu)
 {
+    cpu_set_t cpuset;
     MachineState *ms = MACHINE(qdev_get_machine());
+    MachineClass *mc = MACHINE_GET_CLASS(ms);
+    unsigned affinity = mc->vcpu_affinity[cpu->cpu_index];
 
     cpu->nr_cores = machine_topo_get_cores_per_socket(ms);
     cpu->nr_threads =  ms->smp.threads;
@@ -683,6 +686,12 @@ void qemu_init_vcpu(CPUState *cpu)
     g_assert(cpus_accel != NULL && cpus_accel->create_vcpu_thread != NULL);
     cpus_accel->create_vcpu_thread(cpu);
 
+    if (affinity != -1) {
+        CPU_ZERO(&cpuset);
+        CPU_SET(affinity, &cpuset);
+        pthread_setaffinity_np((cpu->thread)->thread, sizeof(cpu_set_t), &cpuset);
+    }
+
     while (!cpu->created) {
         qemu_cond_wait(&qemu_cpu_cond, &bql);
     }
diff --git a/system/vl.c b/system/vl.c
index 0dbdba6421..bd1dae8322 100644
--- a/system/vl.c
+++ b/system/vl.c
@@ -719,6 +719,22 @@ static void configure_blockdev(BlockdevOptionsQueue *bdo_queue,
 
 }
 
+static QemuOptsList qemu_vcpu_opts = {
+    .name = "vcpu-opts",
+    .implied_opt_name = "vcpunum",
+    .head = QTAILQ_HEAD_INITIALIZER(qemu_vcpu_opts.head),
+    .desc = {
+        {
+            .name = "vcpunum",
+            .type = QEMU_OPT_NUMBER,
+        }, {
+            .name = "affinity",
+            .type = QEMU_OPT_NUMBER,
+        },
+        { /*End of list */ }
+    },
+};
+
 static QemuOptsList qemu_smp_opts = {
     .name = "smp-opts",
     .implied_opt_name = "cpus",
@@ -2790,6 +2806,7 @@ void qemu_init(int argc, char **argv)
     qemu_add_opts(&qemu_accel_opts);
     qemu_add_opts(&qemu_mem_opts);
     qemu_add_opts(&qemu_smp_opts);
+    qemu_add_opts(&qemu_vcpu_opts);
     qemu_add_opts(&qemu_boot_opts);
     qemu_add_opts(&qemu_add_fd_opts);
     qemu_add_opts(&qemu_object_opts);
@@ -3419,6 +3436,12 @@ void qemu_init(int argc, char **argv)
                     exit(1);
                 }
                 break;
+            case QEMU_OPTION_vcpu:
+                if (!qemu_opts_parse_noisily(qemu_find_opts("vcpu-opts"),
+                                             optarg, true)) {
+                    exit(1);
+                }
+                break;
 #ifdef CONFIG_VNC
             case QEMU_OPTION_vnc:
                 vnc_parse(optarg);
@@ -3742,6 +3765,7 @@ void qemu_init(int argc, char **argv)
     qemu_apply_legacy_machine_options(machine_opts_dict);
     qemu_apply_machine_options(machine_opts_dict);
     qobject_unref(machine_opts_dict);
+    parse_vcpu_opts(current_machine);
     phase_advance(PHASE_MACHINE_CREATED);
 
     /*
diff --git a/util/qemu-thread-posix.c b/util/qemu-thread-posix.c
index b2e26e2120..1307860b81 100644
--- a/util/qemu-thread-posix.c
+++ b/util/qemu-thread-posix.c
@@ -506,6 +506,10 @@ static void *qemu_thread_start(void *args)
     void *arg = qemu_thread_args->arg;
     void *r;
 
+#ifdef PRINT_THREADS_IDS
+    qemu_log(" -> PID %d\n", gettid());
+#endif
+
     /* Attempt to set the threads name; note that this is for debug, so
      * we're not going to fail if we can't set it.
      */
@@ -578,6 +582,10 @@ void qemu_thread_create(QemuThread *thread, const char *name,
     qemu_thread_args->start_routine = start_routine;
     qemu_thread_args->arg = arg;
 
+#ifdef PRINT_THREADS_IDS
+    qemu_log("Creating thread '%s'", name);
+#endif
+
     err = pthread_create(&thread->thread, &attr,
                          qemu_thread_start, qemu_thread_args);
 
